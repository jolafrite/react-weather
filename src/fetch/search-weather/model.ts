export interface ISearchWeatherFetchResponse {
  message:	string;
  cod:	number
  count:	number
  list: ISearchWeather[]
}

export interface ISearchWeather {
  id: number;
  name: string;
  coord: {
    lat: number;
    lon: number;
  };
  main: {
    temp: number;
    pressure: number;
    humidity: number;
    temp_min: number;
    temp_max: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  dt: number;
  wind: {
    speed: number;
    deg: number;
  };
  sys: {
    country: string
  };
  rain: {
    "1h": number;
    "3h": number;
  };
  snow: {
    "1h": number;
    "3h": number;
  };
  clouds: {
    all: number;
  };
  weather: {
    id: number;
    main: string;
    description: string;
    icon: string
  }[];
}
