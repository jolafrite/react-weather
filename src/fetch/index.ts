export * from './load-weather/load-weather.fetch';
export * from './search-weather/search-weather.fetch';
