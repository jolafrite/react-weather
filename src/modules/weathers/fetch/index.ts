export * from './load-5-days-weather-forecast/load-5-days-weather-forecast.fetch';
export * from './load-weather/load-weather.fetch';
export * from './search-weather/search-weather.fetch';
