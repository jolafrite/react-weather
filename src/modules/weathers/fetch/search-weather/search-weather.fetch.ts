import Axios from "axios";
import {
  IWeatherMain,
  IWeatherWind,
  IWeatherRain,
  IWeatherSnow,
  IWeatherClouds,
  IWeatherWeather
} from "../../models";

export interface ISearchWeatherFetchResponse {
  message: string;
  cod: string;
  count: number;
  list: {
    id: number;
    name: string;
    coord: {
      lat: number;
      lon: number;
    };
    main: IWeatherMain;
    dt: number;
    wind: IWeatherWind;
    sys: {
      country: string;
    };
    rain?: IWeatherRain | null;
    snow?: IWeatherSnow | null;
    clouds: IWeatherClouds;
    weather: IWeatherWeather[];
  }[];
}

export const searchWeatherFetch = async (query: string) => {
  const { data } = await Axios.request<ISearchWeatherFetchResponse>({
    method: "get",
    url: `${process.env.WEATHER_URL}/find`,
    params: {
      q: query,
      units: "metric",
      appid: "{WEATHER_APPID}"
    }
  });

  return data;
};
