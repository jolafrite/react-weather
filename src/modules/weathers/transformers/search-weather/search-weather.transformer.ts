import { ISearchWeather } from "../../../../fetch/search-weather/model";
import { IWeather } from "../../models";
import { utcToDateTimeString } from "../../../../common/utils/date";

export const searchWeatherListTransformer = (weathers: ISearchWeather[]): IWeather[] => {
  return weathers.map(weather => searchWeatherTransformer(weather));
}

export const searchWeatherTransformer = (weather: ISearchWeather): IWeather => {
  return {
    city: {
      id: weather.id,
      name: weather.name,
      coord: {
        ...weather.coord
      },
      country: weather.sys.country
    },
    cnt: 1,
    list: [
      {
        dt: weather.dt,
        main: {
          ...weather.main
        },
        weather: [
          ...weather.weather,
        ],
        clouds: {
          ...weather.clouds
        },
        wind: {
          ...weather.wind,
        },
        dt_txt: utcToDateTimeString(weather.dt),
      }
    ]
  };
}
