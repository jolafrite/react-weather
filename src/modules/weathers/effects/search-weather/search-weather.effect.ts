import { WeatherSearchPageActions, WeatherApiActions } from '../../actions';
import { Dispatch } from 'redux';
import { searchWeatherFetch } from '../../fetch';
import { searchWeatherFetchResponseTransformer } from '../../transformers';

export const searchPageSearchWeatherEffect = (query: string) => async (dispatch: Dispatch) => {
  dispatch(WeatherSearchPageActions.search(query));

  return await searchWeathersEffect(
    query,
    WeatherSearchPageActions.searchSuccess,
    WeatherSearchPageActions.searchError,
    dispatch
  );
}

const searchWeathersEffect = async (query: string, onSuccess: any, onError: any, dispatch: Dispatch) => {
  try {
    const response = await searchWeatherFetch(query);
    const transformedResponse = searchWeatherFetchResponseTransformer(response);

    dispatch(WeatherApiActions.loadListSuccess(transformedResponse));
    dispatch(onSuccess(transformedResponse));
  } catch(e) {
    dispatch(onError(e));
  }
};
