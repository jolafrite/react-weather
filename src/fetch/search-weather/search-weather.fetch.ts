import Axios from "axios";
import { IWeather } from "../../modules/weathers/models";
import { ISearchWeatherFetchResponse } from "./model";
import { searchWeatherListTransformer } from "../../modules/weathers/transformers";

export const searchWeatherFetch = async (query: string): Promise<IWeather[]> => {
  //qery must be minimum 3 characters long
  if(query.length < 3) {
    return [];
  };

  const { data } = await Axios.request<ISearchWeatherFetchResponse>({
    method: 'get',
    url: `${process.env.WEATHER_URL}/find`,
    params: {
      q: query,
      units: 'metric',
      appid: '{WEATHER_APPID}',
    }
  });

  return searchWeatherListTransformer(data.list);
};
