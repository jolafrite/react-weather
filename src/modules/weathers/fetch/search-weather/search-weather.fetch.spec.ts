import { searchWeatherFetch } from "./search-weather.fetch";
import Axios from "axios";
import { cleanup } from "@testing-library/react";

jest.mock("axios");

describe('fetch: searchWeatherFetch', () => {

  afterEach(cleanup)

  describe('call with a query', () => {
    test('should call the api and return the response list', async () => {
      const axiosRequest = (Axios.request as jest.Mock<any>);

      axiosRequest.mockResolvedValueOnce({
        data: {
          list: 'the response'
        }
      });

      const response = await searchWeatherFetch('paris');

      expect(
        axiosRequest.mock.calls
      ).toMatchSnapshot();

      expect(
        response
      ).toMatchSnapshot()
    });
  });
});
