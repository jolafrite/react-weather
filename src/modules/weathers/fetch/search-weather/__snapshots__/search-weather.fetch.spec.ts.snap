// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch: searchWeatherFetch call with a query less that 3 characters should return an empty array 1`] = `Array []`;

exports[`fetch: searchWeatherFetch call with a query should call the api and return the response list 1`] = `
Array [
  Array [
    Object {
      "method": "get",
      "params": Object {
        "appid": "{WEATHER_APPID}",
        "q": "paris",
        "units": "metric",
      },
      "url": "undefined/find",
    },
  ],
]
`;

exports[`fetch: searchWeatherFetch call with a query should call the api and return the response list 2`] = `"the response"`;
